# ภาษา Java

ภาษา Java เป็นภาษาสำหรับงานทั่วไปที่สนับสนุนแนวคิดการโปรแกรมเชิงวัตถุ (Object-Oriented Programmin [OOP]) พัฒนาขึ้นโดย James Gosling และทีมงานที่ Sun Microsystems และเปิดตัวในปี 1995

โครงการพัฒนาภาษาใหม่ของ Sun Microsystems มีชื่อดั้งเดิมว่า Oak ตามชื่อของต้นไม้ที่อยู่หน้าห้องทำงานของ Gosling แต่เนื่องจากไปซ้ำกับเครื่องหมายการค้าของบริษัทอื่นจึงจำเป็นต้องเปลี่ยน ทีมพัฒนาระดมสมองร่วมกันในห้องประชุมเพื่อหาชื่อใหม่ มีชื่อที่ถูกเสนอออกมามากมาย เช่น DNA, Silk, Ruby, WRL, Lyric, Pepper, NetProse, Neon, WebDancer, WebSpinner และ Java ก่อนจะได้ข้อสรุปเป็นชื่อ Java

ไม่มีใครในทีมจำได้ว่าใครเป็นคนเสนอชื่อ Java และมีอะไรเป็นแรงบันดาลใจ แต่สันนิษฐานว่ามาจากการที่สมาชิกในทีมชอบดื่มกาแฟ และหนึ่งในนั้นก็มีกาแฟชวา (Java) อยู่ในมือ เราจะเห็นได้จากสัญลักษณ์ของภาษา Java ที่มักจะเป็นรูปถ้วยกาแฟที่กรุ่นควัน

รูปแบบของภาษา Java มีรากฐานมาจากภาษา C และ C++ แต่ออกแบบให้สามารถเขียนโปรแกรมที่ทำงานถูกต้องได้ง่ายขึ้นโดยการตัดการทำงานในระดับต่ำบางอย่างออกไป

ภาษา Java ออกแบบมาให้มีจุดเด่นที่สำคัญดังนี้

- เป็นภาษาเชิงวัตถุที่เข้าใจง่ายและใช้งานง่าย เมื่อเทียบกับภาษาอย่าง C++ ในสมัยนั้น ภาษา Java ถือว่าเรียบง่ายกว่ามาก แต่ก็มีรูปแบบภาษาที่ใกล้เคียง C และ C++ ซึ่งทำให้คนที่ใช้ภาษาเหล่านั้นอยู่ก่อนสามารถเรียนรู้ได้ง่ายขึ้น
- Write Once, Run Anywhere โปรแกรมที่เขียนขึ้นและคอมไพล์แล้วสามารถนำไปรันที่ระบบไหนก็ได้ ถือเป็นคติพจน์ของ Java ที่ได้รับการโฆษณาอย่างมากในช่วงแรก ในทางปฏิบัติก็มีข้อจำกัดเฉพาะระบบที่ทำให้การใช้งานในระบบที่ต่างกันแตกต่างกันออกไปได้บ้าง
- มีการจัดการหน่วยความจำด้วยตัวเก็บขยะ (garbage collector) ภาษาอย่าง C/C++ ผู้เขียนโปรแกรมต้องจัดการทำลายข้อมูลที่สร้างขึ้นมาหลังจากเลิกใช้แล้วด้วยตัวเอง ซึ่งเป็นจุดที่มักจะสร้างความผิดพลาดในการทำงานของโปรแกรมได้มากที่สุดจุดหนึ่ง ภาษา Java ใช้ตัวเก็บขยะจัดการทำลายข้อมูลที่ไม่ใช้แล้วโดยอัตโนมัติ ทำให้ผู้เขียนโปรแกรมไม่ต้องทำเอง ลดความเสี่ยงต่อความผิดพลาดลงได้อีกจุดหนึ่ง
- ตัวแปลภาษา ระบบสนับสนุน เครื่องมือต่าง ๆ ที่ใช้พัฒนาโปรแกรมถูกจัดมาให้โดยไม่ต้องเสียค่าใช้จ่าย เทียบกับสมัยนั้นที่เครื่องมือพัฒนาโปรแกรมดี ๆ บนบางระบบมีราคาสูง

แน่นอนว่าถ้ามองเปรียบเทียบกับปัจจุบัน จุดเด่นเหล่านั้นไม่ได้เด่นอีกต่อไปแล้ว เพราะภาษาสมัยใหม่จำนวนมากก็มีคุณสมบัติเหล่านั้นเหมือนกัน แต่เมื่อเรามองย้อนไปเมื่อปี 1995 การที่ Java เปิดตัวมาพร้อมกับคุณสมบัติทั้งหมดนั้นทำให้ Java ได้รับความนิยมอย่างล้นหลาม และยังคงเป็นหนึ่งในภาษาที่ได้รับความนิยมที่สุดในปัจจุบันอยู่

ในช่วงปี 2009-2010 Oracle ได้เข้าซื้อกิจการของ Sun Microsystems และพัฒนา Java ต่อมาจนถึงปัจจุบัน

Oracle ได้จัดชุดพัฒนาโปรแกรมด้วยภาษา Java ไว้หลายรูปแบบ โดยรูปแบบหลัก ๆ มีดังนี้

- Java SE (Java Platform, Standard Edition) สำหรับการพัฒนาซอฟต์แวร์ทั่วไป
- Java EE (Java Platform, Enterprise Edition) เป็นส่วนขยายจาก Java SE โดยเพิ่มข้อกำหนดในส่วนของการพัฒนาระบบเว็บ การประมวลผลแบบกระจาย และเครื่องมืออื่น ๆ ที่ใช้ในการพัฒนาซอฟต์แวร์ระดับองค์กร
- Java ME (Java Platform, Micro Edition) เน้นที่การพัฒนาซอฟต์แวร์สำหรับอุปกรณ์ฝังตัวและอุปกรณ์เคลื่อนที่ เช่น ไมโครคอนโทรลเลอร์ โทรศัพท์มือถือ เป็นต้น ในทางปฏิบัติ ผู้พัฒนาฮาร์ดแวร์มักจะพัฒนาระบบสนับสนุนของตัวเองซึ่งมี API ที่แตกต่างกันไปขึ้นมาด้วย ประกอบกับการมาของระบบปฏิบัติการ Android พร้อมเครื่องมือพัฒนาของตัวเอง ทำให้ความนิยม Java ME ลดลงไปอย่างมาก

ภาษา Java ผ่านประวัติศาสตร์การใช้งานมาอย่างยาวนาน เทคโนโลยีเสริมที่พัฒนาจากภายนอกทำให้ความจำเป็นที่จะต้องใช้ Java EE และ Java ME ลดลงไปมาก เราสามารถใช้ Java SE ในการทำทุกอย่างที่ Java EE และ Java ME ทำได้โดยใช้ไลบรารีหรือเฟรมเวิร์กอื่น ๆ ช่วย

เราจะใช้ Java SE ในวิชานี้

## การติดตั้ง Java Development Kit

Java Development Kit (JDK) คือชุดพัฒนาโปรแกรมด้วยภาษา Java ซึ่งจะประกอบด้วย Java Virtual Machine (JVM) ที่ใช้ในการรันโปรแกรม ตัวไลบรารีต่าง ๆ และเครื่องมืออื่น ๆ ที่ช่วยในการพัฒนาโปรแกรม

ดาวน์โหลด JDK ได้ที่ http://www.oracle.com/technetwork/java/javase/downloads/index.html

เมื่อติดตั้งเสร็จแล้ว ลองทดสอบด้วยการพิมพ์คำสั่งต่อไปนี้ที่ Command Prompt (Windows), Terminal (Mac OSX) หรือ Shell (Linux)

`javac`

ถ้ารันได้ตามปกติแสดงว่าการติดตั้งเรียบร้อยดี ถ้าไม่พบ `javac` อาจจะเป็นเพราะต้องตั้งค่า `PATH` ให้เรียบร้อยก่อน บนระบบปฏิบัติการ Windows จะอยู่ที่ "C:\Program Files\Java\jdk1.x.x_xxx\bin" โดย x ในที่นี้คือรุ่นของ JDK ที่ติดตั้ง

นอกจาก JDK แล้วเรายังต้องมีเครื่องมือสำหรับเขียนโปรแกรมด้วย ในเบื้องต้นเราจะใช้ Atom ซึ่งดาวน์โหลดได้ที่ https://atom.io

## โปรแกรมแรกด้วยภาษา Java

ใช้ Atom เขียนโค้ดต่อไปนี้ แล้วบันทึกโดยตั้งชื่อไฟล์นี้ว่า Welcome.java

In [1]:
public class Welcome {
    public static void main(String[] args) {
        System.out.println("Welcome to the Java World!");
    }
}

### การคอมไพล์

การตั้งชื่อไฟล์โปรแกรมเราจะตั้งตามชื่อของคลาส (class) จากโค้ดด้านบนนี้ คลาสมีชื่อว่า `Welcome` เราจึงตั้งชื่อไฟล์ว่า Welcome.java

โปรแกรมที่เขียนด้วยภาษา Java ไม่สามารถรันได้ทันที ต้องผ่านขั้นตอนการคอมไพล์ก่อนเพื่อเปลี่ยนโค้ดภาษา Java ให้เป็นรหัสคำสั่งที่ JVM สามารถทำงานได้

เราคอมไพล์โปรแกรมด้วยคำสั่งต่อไปนี้

`javac Welcome.java`

`javac` เป็นชื่อของคอมไพเลอร์ภาษา Java ซึ่งเราใช้ในการคอมไพล์โค้ดโปรแกรมให้เป็นสิ่งที่เราเรียกว่าไบต์โค้ด (bytecode) ซึ่งเป็นชุดคำสั่งของ JVM

จากการคอมไพล์ข้างต้น เราจะได้ผลลัพธ์เป็นไฟล์ชื่อ Welcome.class

### การรัน

ไฟล์ที่มีนามสกุล .class จะอยู่ในรูปของไบต์โค้ดและสามารถรันได้ทันที

เรารันโปรแกรมนี้ด้วยคำสั่ง

`java Welcome`

ซึ่งจะได้ผลลัพธ์เป็นข้อความ

`Welcome to the Java World!`

สังเกตว่าการสั่งรันโปรแกรมด้วยคำสั่ง `java` เราไม่จำเป็นต้องระบุนามสกุลของไฟล์ (ในที่นี้คือ .class) เนื่องจากว่าสิ่งที่ตามหลังคำสั่ง `java` ไม่ใช่ชื่อไฟล์ แต่เป็นชื่อของคลาสที่ต้องการรัน

คำสั่ง `java` เป็นการโหลดไฟล์ .class เข้าไปสู่ JVM หรือ Java Virtual Machine เพื่อให้โปรแกรมเริ่มต้นทำงาน JVM จะทำตัวเป็นเหมือนเครื่องคอมพิวเตอร์เสมือนที่รันอยู่บนเครื่องคอมพิวเตอร์จริง ๆ อีกที แนวคิด Write Once, Run Anywhere เกิดขึ้นได้ด้วยหลักการนี้ เนื่องจากเครื่องคอมพิวเตอร์จริง ๆ แต่ละระบบก็จะมีความแตกต่างกัน โปรแกรมที่คอมไพล์สำหรับระบบหนึ่งจะไม่สามารถไปทำงานบนอีกระบบได้ แต่ Java แก้ปัญหานี้โดยการให้มี JVM ซึ่งเป็นซอฟต์แวร์จำลองระบบคอมพิวเตอร์เสมือน โปรแกรมภาษา Java จะถูกคอมไพล์มาเป็นรหัสคำสั่งของ JVM ซึ่งจะทำให้โปรแกรมนี้สามารถรันได้ทุกที่ทุกระบบที่มี JVM อยู่

### คำอธิบายโปรแกรม

ส่วนที่เป็นการทำงานจริงของโปรแกรมนี้คือ

```java
System.out.println("Welcome to the Java World!");
```

ซึ่งเป็นคำสั่งให้แสดงผลข้อความตามที่ระบุออกทางจอภาพแล้วขึ้นบรรทัดใหม่

นอกเหนือจากบรรทัดนี้ ส่วนอื่น ๆ เป็นส่วนที่จำเป็นต้องมีตามข้อกำหนดของโปรแกรมที่เขียนด้วยภาษา Java

ข้อแรกคือ โค้ดโปรแกรมจะต้องอยู่ในคลาส (class) ซึ่งเป็นหน่วยของโปรแกรมหน่วยหนึ่งที่ประกอบด้วยข้อมูลและโค้ดที่จัดการกับข้อมูลเหล่านั้น เรายังไม่ได้ให้ความสำคัญกับคลาสในตอนนี้

ข้อที่สองคือ โปรแกรมจะเริ่มทำงานที่เมทอด (method) ที่ชื่อ `main` ที่อยู่ภายในคลาส ถ้าเปรียบเทียบกับภาษาอื่น เมทอดก็คือฟังก์ชันที่อยู่ภายในคลาส เราสามารถจะมีเมทอดจำนวนเท่าใดก็ได้ในคลาสหนึ่ง ๆ แต่การทำงานของโปรแกรมจะเริ่มต้นที่เมทอด `main` เท่านั้น

เพราะฉะนั้น โปรแกรม Java ขั้นต่ำที่รันได้จะต้องมีคลาส และมีเมทอด `main` โดยคลาสจะใช้ชื่ออะไรก็ได้ตามหลักการตั้งชื่อ

ในเบื้องต้นขอให้ใช้โครงสร้างนี้ในการเขียนโปรแกรมไปก่อน เราจะกลับมาพูดถึงรายละเอียดของแต่ละส่วนอีกครั้งในบทที่เหมาะสม

```java
public class ClassName {
    public static void main(String[] args) {
        // Your code here
    }
}
```

`ClassName` ให้เราตั้งเป็นชื่อคลาสตามต้องการ และเขียนโค้ดอยู่ภายในเมทอด `main`

#### การเขียนคอมเมนต์

เราสามารถระบุส่วนที่เป็นคอมเมนต์หรือหมายเหตุในโปรแกรมได้ 3 รูปแบบดังนี้

```java
/* This is a block comment. Use this when you have long comments,
   usually spanning multiple lines. */

// And this is a single-line comment. It ends at the end of the line.

/**
 * This is a javadoc comment. We can use 'javadoc' tool to automatically
 * generate documentation.
 */
```

### ตัวแปรและชนิดข้อมูล

ตัวแปรในภาษา Java จะต้องถูกประกาศและกำหนดชนิดข้อมูลก่อนการใช้งาน เมื่อตัวแปรถูกกำหนดชนิดแล้วจะไม่สามารถเปลี่ยนแปลงได้

ชนิดข้อมูลในภาษา Java จะมี 2 ประเภท ได้แก่ชนิดพื้นฐาน (primitive type) และชนิดตัวอ้างอิง (reference type)

ชนิดข้อมูลพื้นฐานมีดังต่อไปนี้

| ชนิดข้อมูล | ลักษณะ    | ตัวอย่าง | ขนาด (บิต) | ช่วงของค่า |
|---------|----------|--------|----------:|----------|
| boolean | ตรรกะ    | false, true | 1 |  |
| char    | อักขระ    | 'a', 'A', ' ', '0', '\u0040' | 16 | \u0000 ... \uFFFF |
| byte    | จำนวนเต็ม | 0, 1, -1, -127, 50 | 8 | 2<sup>8</sup> -128 .. 127 |
| short   | จำนวนเต็ม | 0, 1, -1, 1000, 2_000 | 16 | -32,768 .. 32,767 |
| int     | จำนวนเต็ม | 0, 1, -1, 200_000 | 32 | -2,147,483,648 .. 2,147,483,647 |
| long    | จำนวนเต็ม | 0L, 1L, -1L, 2000L | 64 | -9,223,372,036,854,775,808 .. 9,223,372,036,854,775,807 |
| float   | ค่าทศนิยม  | -1.0f, 0.49f, 3E8f | 32 | ประมาณ ±1.4×10<sup>-45</sup> .. ±3.4×10<sup>38</sup>, นัยสำคัญ 7-8 หลัก |
| double  | ค่าทศนิยม  | -1.0, 0.49, 3E8 | 64 | ประมาณ ±4.9×10<sup>-324</sup> .. ±1.7×10<sup>308</sup>, นัยสำคัญ 15-16 หลัก |

ข้อมูลชนิดตัวอ้างอิงคือข้อมูลที่มีชนิดเป็นคลาส อินเทอร์เฟซ อาร์เรย์ หรือ `enum` ข้อมูลชนิด `String` ในภาษา Java มีชนิดเป็นตัวอ้างอิง เพราะ `String` เป็นคลาส

ตัวแปรที่ประกาศชนิดเป็นชนิดพื้นฐานจะเก็บข้อมูลตามชนิดนั้นไว้ที่ตัวแปรโดยตรง แต่ตัวแปรที่ประกาศชนิดเป็นตัวอ้างอิงจะไม่ได้เก็บตัวข้อมูลโดยตรง แต่จะเก็บตัวอ้างอิงไปยังข้อมูล เราจะพูดถึงตัวอ้างอิงอีกครั้งในบทเรื่องคลาสและอ็อบเจกต์

#### การระบุชนิดของค่าตัวเลข

จากตารางเราจะเห็นว่าค่าตัวเลขชนิด `long` จะมีการระบุ `L` ตามท้ายเพื่อบอกว่าเป็นตัวเลขชนิด `long` และชนิด `float` ก็จะมีการระบุ `f` ตามหลังเช่นกัน

ถ้าไม่ระบุ เลขจำนวนเต็มจะถือเป็นชนิด `int` และเลขทศนิยมจะถึอเป็นชนิด `double`

#### รูปแบบการเขียนจำนวนเต็ม

ค่าจำนวนเต็มสามารถเขียนได้ในรูปฐาน 10 และในรูปฐานอื่น ๆ ดังนี้

| ฐาน                  | รูปแบบ             | ตัวอย่าง |
|----------------------|-------------------|--------|
| ฐาน 10 (decimal)     | ไม่ขึ้นต้นด้วย 0       | 1234, -15, 1_500 |
| ฐาน 2 (binary)       | ขึ้นต้นด้วย 0b หรือ 0B | 0b101, 0B1110011 -0b10_1100 |
| ฐาน 8 (octal)        | ขึ้นต้นด้วย 0         | 01_234, -0723, 0755 |
| ฐาน 16 (hexadecimal) | ขึ้นต้นด้วย 0x หรือ 0X | 0xAB_CDEF, -0x7F, 0x12ABCF |

เราสามารถใช้เครื่องหมาย `_` (underscore) ในการแบ่งหลักของตัวเลขได้ การแบ่งหลักไม่มีข้อบังคับว่าจะต้องแบ่งเป็นชุดละกี่หลัก ควรใช้เพื่อเพิ่มความชัดเจนในการแสดงตัวเลขตามความเหมาะสม

#### รูปแบบการเขียนจำนวนทศนิยม

เราสามารถเขียนเลขทศนิยมตามรูปปกติ หรือรูปยกกำลังก็ได้ เช่น -25.24, 0.41, .25, 1.5E-7, -2.54E12 เป็นต้น รูปยกกำลังจะประกอบด้วยตัวคูณซึ่งก็คือตัวเลขก่อนตัว 'E' และกำลังของ 10 ซึ่งก็คือเลขหลังตัว 'E' เช่น 1.54E-7 ก็จะหมายถึง 1.5×10<sup>-7</sup>

#### ข้อมูลชนิดตัวเลขกับการล้นและการน้อยเกินเก็บ

การเก็บข้อมูลในคอมพิวเตอร์มีข้อจำกัดที่ขนาดของข้อมูล จากตารางชนิดข้อมูลพื้นฐานเราจะเห็นว่าตัวเลขแต่ละชนิดมีการระบุช่วงของค่าที่เป็นไปได้อยู่ด้วย เช่น `int` มีการระบุไว้เป็นช่วงตั้งแต่ -2,147,483,648 ถึง 2,147,483,647 และ `byte` มีค่าได้ตั้งแต่ -128 ถึง 127 เป็นต้น

เราลองศึกษาพฤติกรรมของตัวเลขเมื่อเราพยายามเพิ่มค่าไปมากกว่าค่าสูงสุดที่เป็นไปได้จากส่วนของโค้ดต่อไปนี้

In [8]:
int value = Integer.MAX_VALUE;
System.out.printf("Now: value = %,d%n", value);
value = value + 1;
System.out.printf("After increment: value = %,d%n", value);
value = value - 1;
System.out.printf("After decrement: value = %,d%n", value);

Now: value = 2,147,483,647
After increment: value = -2,147,483,648
After decrement: value = 2,147,483,647


> **หมายเหตุ** ส่วนของโค้ดด้านบนนั้นไม่สามารถคอมไพล์ได้โดยตรง ต้องเขียนอยู่ในโครงของคลาสและเมทอด `main` จึงจะนำไปคอมไพล์ได้ แต่เพื่อความกระชับ ในบางครั้งเราจะละโครงของคลาสและเมทอด `main` ในเอกสาร เวลาที่ทดสอบจริงต้องเขียนโครงเพิ่มเข้ามาด้วย

เรากำหนดให้ตัวแปร `value` มีค่าเป็น `Integer.MAX_VALUE` ซึ่งเป็นค่าสูงสุดที่ตัวแปรชนิด `int` จะเก็บได้ เราพบว่าเมื่อเพิ่มค่าจากค่าสูงสุดไปอีก 1 ค่า ผลลัพธ์ที่ได้คือ `value` จะมาอยู่ที่ค่าต่ำสุดที่เป็นไปได้ (-2,147,483,648) และเมื่อลดค่าจากค่าต่ำสุดไป 1 ค่า ค่าของ `value` ก็จะกลับมาอยู่ที่ค่าสูงสุดอีก

พฤติกรรมที่เราสังเกตได้นี้คือการล้น (overflow) ของค่า ตัวเลขจำนวนเต็มในคอมพิวเตอร์โดยทั่วไปแล้วจะจัดเก็บในรูปแบบที่เรียกว่าส่วนเติมเต็มของสอง (two's complement) ซึ่งในรูปแบบนี้ หากค่าตัวเลขมีค่าเป็นบวกสูงสุดแล้ว การเพิ่มค่าไปอีกจะทำให้ค่าไปอยู่ทางฝั่งเลขลบที่ต่ำที่สุด ในทางกลับกัน หากค่าตัวเลขมีค่าเป็นค่าลบต่ำที่สุดแล้ว การลดค่าไปอีกจะทำให้ค่าไปอยู่ฝั่งบวกที่สูงที่สุดแทน

ภาษา Java ไม่มีการตรวจสอบการล้นของจำนวนเต็มโดยอัตโนมัติ เป็นหน้าที่ของโปรแกรมเมอร์ที่จะต้องเลือกใช้ชนิดข้อมูลที่ใหญ่เพียงพอสำหรับข้อมูลที่จะจัดเก็บ และระมัดระวังในการใช้งานไม่ให้เกิดการล้นได้

เรามาดูในส่วนของเลขทศนิยมกันบ้าง

In [13]:
System.out.println("Floating-point overflow test:");
System.out.println("Max double: " + Double.MAX_VALUE);
System.out.println("Added with a large value: " + (Double.MAX_VALUE + 1_000_000.0));
System.out.println("Adding very large values: " + (1.5E308 + 1.5E308));

Floating-point overflow test:
Max double: 1.7976931348623157E308
Added with a large value: 1.7976931348623157E308
Adding very large values: Infinity


เราจะเห็นว่าค่าที่มีขนาดมากที่สุดของจำนวนทศนิยมชนิด `double` คือ 1.7976931348623157×10<sup>308</sup> เมื่อเราทดสอบโดยการบวกค่าขนาดใหญ่เพิ่มไปจากค่าที่มากที่สุด พบว่าเราจะได้ผลลัพธ์เป็นค่ามากที่สุดเหมือนเดิม แต่เมื่อเราเอาค่าที่มีขนาดใหญ่มาก ใกล้เคียงกับค่ามากที่สุดสองตัวมาบวกกัน เราจะได้ผลลัพธ์เป็น `Infinity`

การเก็บเลขทศนิยมในคอมพิวเตอร์เป็นการเก็บค่าประมาณ โดยมีจำนวนตัวเลขนัยสำคัญ (significant digit) ที่จำกัด เช่น ของชนิด `float` มีตัวเลขนัยสำคัญประมาณ 7-8 หลัก และ `double` ประมาณ 15-16 หลัก

เมื่อเลขเข้าใกล้ค่าสูงสุด ความละเอียดในหลักต่ำ ๆ ก็จะลดลง การกระทำกับเลขที่มีขนาดเล็กกว่ามาก ๆ อาจจะไม่ส่งผลใด ๆ เลยเนื่องจากการปัดค่าเข้าหาเลขที่สูงกว่าเหมือนเดิม แต่ถ้าเอาเลขที่มีขนาดใหญ่มาก ๆ ใกล้ค่าสูงสุดมาบวกกัน ผลที่ได้จะมีขนาดใหญ่มากเกินกว่าจะเก็บได้ และทำให้เกิดการล้นขึ้น ซึ่งในกรณีของ Java จะมีการตรวจสอบการล้นของเลขทศนิยม และกำหนดผลลัพธ์ให้เป็น `Infinity`

พฤติกรรมนี้เหมือนกันทั้งเลขบวกและเลขลบที่มีขนาดใหญ่มาก ๆ

ลองพิจารณาเลขทศนิยมที่มีขนาดเล็กมาก ๆ ดูบ้าง

In [20]:
System.out.println("Floating-point underflow test:");
System.out.println("Min double: " + Double.MIN_VALUE);
System.out.println("Subtracting a small value: " + (Double.MIN_VALUE - 1.0E-50));
System.out.println("Multiplying very small values: " + (4.0E-320 * 4.0E-320));

Floating-point underflow test:
Min double: 4.9E-324
Subtracting a small value: -1.0E-50
Multiplying very small values: 0.0


ค่าที่เล็กที่สุดของ `double` คือ 4.9E-324 ถ้าเราลองเอาค่าขนาดเล็ก ๆ อีกค่ามาลบออก เช่น ในตัวอย่างนี้คือ 1.0E-50 จะพบว่าผลลัพธ์กลายเป็น -1.0E-50 ทั้งนี้เพราะว่าค่าที่เล็กที่สุดเป็นค่าที่เล็กกว่าค่าที่เอามาลบมาก ๆ เมื่อลบกันจึงเกิดการปัดค่าเนื่องจากผลลัพธ์มีจำนวนตัวเลขนัยสำคัญเกินกว่าที่ `double` จะเก็บได้

แต่ถ้าเอาเลขที่เล็กมาก ๆ มากระทำต่อกันเช่นมาคูณกัน ในตัวอย่างนี้เราจะพบว่าผลลัพธ์ที่ได้เป็น 0.0 เพราะว่าค่าผลลัพธ์เล็กเกินกว่าที่ `double` จะเก็บได้ ทำให้เกิดการปัดไปเป็น 0.0 ลักษณะการเกิดค่าที่เล็กเกินกว่าจะเก็บได้นี้เราเรียกว่าการน้อยเกินเก็บ (underflow) ซึ่งจะเกิดกับเลขทศนิยม

เลขทศนิยมใน Java ทั้งชนิด `float` และ `double` มีรูปแบบการจัดเก็บและกฎเกณฑ์การจัดการกับค่าต่าง ๆ ตามมาตรฐาน [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754) ซึ่งสามารถศึกษาดูได้หากสนใจ

#### การเปรียบเทียบค่าทศนิยม

จำนวนทศนิยมในคอมพิวเตอร์ที่จัดเก็บแบบ floating-point เป็นเพียงค่าประมาณของค่าจริง ๆ ในระบบเลขฐาน 10 เลขบางจำนวนจะไม่สามารถแทนด้วยตัวเลขแบบ floating-point ได้ตรงตามค่าเนื่องจากเลข floating-point จะจัดเก็บทศนิยมโดยอ้างอิงเลขฐาน 2 เช่น ในขณะที่เลข 0.5 ในฐาน 10 จะแปลงเป็นเลขฐาน 2 ได้เป็น 0.1 ซึ่งสามารถจัดเก็บได้ตามปกติ เลข 0.1 ในฐาน 10 เมื่อแปลงเป็นเลขฐาน 2 เรากลับได้ค่าเป็น 0.0001100110011001100... เป็นทศนิยมซ้ำไม่สิ้นสุด ซึ่งแน่นอนว่าไม่สามารถจัดเก็บได้ทั้งหมด ดังนั้นเลข 0.1 ในฐาน 10 จึงไม่สามารถจัดเก็บได้อย่างแม่นยำด้วยระบบเลข floating-point ฐาน 2

ลองพิจารณาอีกตัวอย่าง

In [27]:
System.out.println(1.0 - 0.8);

0.19999999999999996


การลบเลขซึ่งน่าจะได้ผลลัพธ์เป็น 0.2 กลับได้เป็นค่าใกล้เคียง แสดงให้เห็นถึงความไม่แม่นยำของเลขทศนิยมแบบ floating-point ซึ่งกรณีเหล่านี้ แม้ว่าเราจะใช้จำนวนบิตมากเท่าใดในการจัดเก็บ เราก็ยังไม่สามารถเก็บผลลัพธ์ได้อย่างแม่นยำทุกกรณีอยู่ดี

ความไม่แม่นยำนี้ทำให้เราควรระวังในการใช้งานเลขทศนิยม โดยเฉพาะเมื่อใช้ในเงื่อนไข เช่น

In [29]:
double mixedSubstance = 0.8;
double residue = 0.2;

if (1.0 - mixedSubstance == residue) {
    System.out.println("Expected residue achieved");
} else {
    System.out.println("Unexpected state reached");
}

Unexpected state reached


เราคาดหวังให้เงื่อนไขเป็นจริง เพราะ 1.0 - 0.8 ควรจะได้เท่ากับ 0.2 แต่ผลลัพธ์กลับเข้ากรณีเงื่อนไขเป็นเท็จ เหตุผลก็ด้วยความไม่แม่นยำที่เรากล่าวถึงก่อนหน้านี้

แนวทางจัดการกับปัญหาลักษณะนี้คือ ในกรณีที่เราต้องการเปรียบเทียบค่าทศนิยม อย่าเปรียบเทียบโดยการเปรียบเทียบว่าเท่ากัน (ด้วยเครื่องหมาย `==`) แต่ให้เปรียบเทียบว่าค่ามีความแตกต่างกันไม่เกินค่าส่วนต่างเล็ก ๆ ที่ยอมรับได้

ในกรณีตัวอย่างด้านบนนี้ เราสามารถเขียนใหม่ได้ดังนี้

In [33]:
double mixedSubstance = 0.8;
double residue = 0.2;
double permittedError = 1.0E-12;

if (Math.abs((1.0 - mixedSubstance) - residue) < permittedError) {
    System.out.println("Expected residue achieved");
} else {
    System.out.println("Unexpected state reached");
}

Expected residue achieved


แทนที่เราจะเปรียบเทียบค่ากันตรง ๆ เราจะใช้วิธีการนำค่าที่ต้องการเปรียบเทียบมาลบกัน แล้วดูว่าค่าสัมบูรณ์ (absolute) ของผลต่างน้อยกว่าค่าความคลาดเคลื่อนที่ยอมรับได้หรือไม่ ในที่นี้ก็คือ `permittedError` ซึ่งมีค่าเป็น 1.0E-12 หรือ 0.000000000001 ซึ่งวิธีนี้ทำให้ได้ผลลัพธ์ที่ถูกต้องตามที่เราตั้งใจ

อย่างไรก็ดี เลขทศนิยมแบบ floating-point ดีพอสำหรับสำหรับการใช้งานทั่ว ๆ ไปที่เราไม่ได้ต้องการความแม่นยำอย่างยิ่งยวด  และเมื่อเรานำค่ามาแสดงผลแบบจำกัดตำแหน่งทศนิยม ฟังก์ชันแสดงผลก็จะปัดเศษให้ตามจำนวนตำแหน่งทศนิยมที่ต้องการ ซึ่งก็ทำให้เรารู้สึกเหมือนได้ผลลัพธ์ที่ถูกต้องได้

In [28]:
System.out.printf("%.2f", 1.0 - 0.8);

0.20

ถ้าเราต้องการความแม่นยำในการจัดเก็บเลขทศนิยมฐาน 10 เช่น ในระบบการเงินหรือธนาคาร ซึ่งความแม่นยำเป็นสิ่งที่จำเป็นมาก ควรใช้คลาส `BigDecimal` ในการจัดเก็บค่าแทน `float` หรือ `double` เนื่องจาก `BigDecimal` จะจัดเก็บจำนวนทศนิยมในรูปแบบที่อิงเลขฐาน 10

ทั้งนี้ ถ้าเราไม่ได้ต้องการความแม่นยำระดับนั้น แต่ต้องการความเร็วในการประมวลผล ข้อมูลชนิด `float` หรือ `double` จะมีประสิทธิภาพสูงกว่า `BigDecimal`

### ข้อมูลชนิดตัวอักษร

Java จัดเก็บข้อมูลชนิดตัวอักษร (`char`) ในแบบ UTF-16 ซึ่งเป็นรูปแบบหนึ่งของรหัสอักขระ Unicode ตัวอักษรปกติใน UTF-16 จะมีขนาด 16 บิตหรือ 2 ไบต์

การเขียนข้อมูลชนิดตัวอักษร เราจะใช้เครื่องหมาย `'` (single quote) ครอบตัวอักษรเอาไว้ เช่น `'a'`, `'B'`, `'!'`, `' '` หรือ `'3'`

นอกจากนี้ เรายังสามารถระบุเป็นรหัส Unicode ได้ด้วย โดยระบุเป็นเลขฐาน 16 ในรูปแบบ `'\uXXXX'` เช่น `'\u0040'` ซึ่งเท่ากับ `'@'`

ในกรณีของตัวอักษรพิเศษ เช่น `'` หรือ `\` เราสามารถใช้ `\` (escape character) เพื่อระบุได้ ซึ่งจะได้เป็น `'\''` และ `'\\'`

### ข้อมูลชนิดสตริง

สตริงหรือข้อความจะจัดเก็บอยู่ในรูปของคลาส `String`

ตัวข้อมูลชนิดสตริงสามารถเขียนได้โดยใช้เครื่องหมาย `"` (double quote) ครอบข้อความเอาไว้

เราสามารถนำสตริงมาต่อกันด้วยเครื่องหมาย `+` ได้ และ Java ยังอนุญาตให้เราต่อสตริงกับข้อมูลชนิดอื่นด้วยเครื่องหมาย `+` ได้ด้วย โดยข้อมูลชนิดอื่นจะถูกแปลงชนิดเป็นสตริงก่อนแล้วจึงนำมาต่อกัน

In [39]:
String messageA = "We've lost ";
String messageB = " match already!";
String messageC = " matches already!";

String messageD = messageA + 3 + messageC;
String messageE = 4.25 + " is a small point.";

System.out.println(messageA + "a" + messageB);
System.out.println(messageD);
System.out.println(messageE);

We've lost a match already!
We've lost 3 matches already!
4.25 is a small point.


ควรระมัดระวังการเชื่อมสตริงกับตัวเลขที่ต้องการบวกกันด้วย

In [42]:
System.out.println("1 + 2 = " + 1 + 2);
System.out.println("1 + 2 = " + (1 + 2));

1 + 2 = 12
1 + 2 = 3


### การแปลงชนิดข้อมูล

เมื่อเรากำหนดค่าชนิดหนึ่งให้กับตัวแปรที่มีชนิดต่างกัน Java จะพยายามแปลงชนิดข้อมูลให้โดยอัตโนมัติ โดยมีเงื่อนไขว่าชนิดปลายทางจะต้องกว้างกว่าชนิดต้นทาง (widening conversion) คำว่ากว้างกว่าในที่นี้คือมีช่วงของค่าที่เป็นไปได้กว้างกว่าช่วงของชนิดต้นทาง ซึ่งสำหรับข้อมูลชนิดตัวเลขแล้วมีลำดับจากแคบที่สุดไปกว้างที่สุดดังนี้

`byte` -> `short` -> `int` -> `long` -> `float` -> `double`

ทั้งนี้ การแปลงข้อมูลจากชนิด `int` ไปยัง `float` หรือจากชนิด `long` ไปยัง `float` หรือ `double` เป็นการแปลงที่อาจเกิดการสูญเสียความแม่นยำของค่าได้ กล่าวคือ ค่าที่ได้หลังจากการแปลงชนิดอาจไม่เท่ากันกับค่าเดิมเนื่องด้วยข้อจำกัดในด้านจำนวนตัวเลขนัยสำคัญของเลขทศนิยมแบบ floating-point

ถึงแม้การแปลงจากชนิด `long` ไปยัง `double` อาจเกิดการสูญเสียความแม่นยำได้ แต่การแปลงจากชนิด `int` ไปยัง `double` เป็นการแปลงที่ไม่เกิดการสูญเสีย เพราะ `double` มีขนาดของข้อมูลกว้างพอที่จะเก็บค่าของ `int` ได้ทั้งหมด

หากชนิดปลายทางแคบกว่าชนิดต้นทาง Java จะไม่แปลงชนิดให้โดยอัตโนมัติ เนื่องจากไม่สามารถรับรองได้ว่าชนิดปลายทางจะรองรับค่าของชนิดต้นทางได้ ในกรณีนี้เราจำเป็นต้องระบุให้แปลงชนิดเอง มิฉะนั้นคอมไพเลอร์จะแจ้งเป็นข้อผิดพลาดและไม่คอมไพล์โค้ดต่อ

การระบุให้แปลงชนิด (type casting) ทำได้โดยรูปแบบ `(type)value` เช่น

```java
int n = (int)2.5;
```

การแปลงชนิดจาก `float` หรือ `double` ไปยังชนิดจำนวนเต็มจะมีการตัดทศนิยมทิ้ง

การสั่งให้แปลงชนิดจากชนิดที่กว้างกว่าไปยังชนิดที่แคบกว่า (narrowing conversion) อาจทำให้ได้ผลลัพธ์ที่ไม่ถูกต้องได้ถ้าค่าต้นทางใหญ่เกินกว่าที่ปลายทางจะเก็บได้ การใช้งานจึงควรกระทำด้วยความระมัดระวัง

#### ตัวดำเนินการทางคณิตศาสตร์กับการแปลงชนิด

ตัวดำเนินการทางคณิตศาตร์เบื้องต้นของ Java ประกอบด้วย `+`, `-`, `*`, `/` และ `%` (modulo)

การดำเนินการ เช่น การ `+` ระหว่างข้อมูลชนิด `int` จะได้ผลลัพธ์เป็นข้อมูลชนิด `int` เช่นเดียวกันกับ `long`, `float` และ `double` ถ้าการดำเนินการเป็นระหว่างชนิดเดียวกันใน 4 ชนิดนั้น ผลลัพธ์ที่ได้จะเป็นชนิดเดิม

แต่ถ้าชนิดข้อมูลต่างกัน คอมไพเลอร์จะแปลงชนิดข้อมูลให้เหมือนกันก่อน โดยจะแปลงไปหาชนิดที่กว้างกว่า ซึ่งผลลัพธ์ก็จะเป็นชนิดที่กว้างกว่าด้วย

ทั้งนี้มีข้อยกเว้นสำหรับชนิด `byte` และ `short` ซึ่งจะแปลงไปหาชนิด `int` หรือชนิดที่กว้างกว่าก่อนเสมอถึงแม้จะดำเนินการระหว่างชนิดเดียวกันหรือไม่ก็ตาม

ตัวอย่างต่อไปนี้แสดงให้เห็นถึงการแปลงชนิดแบบต่าง ๆ บางกรณีที่ไม่สามารถแปลงได้จะคอมเมนต์ไว้ในตัวอย่างว่า "`// ERROR`"

In [ ]:
byte a = 1;
short b = 2;
int c = 3;
int d = 4L; // ERROR: long -> int
long e = 5; // OK: int -> long
long f = 6L;
float x = 7.0; // ERROR: double -> float
float y = 8.0f;
double z = 9.0;

a = a + a; // ERROR: result is an int, int -> byte
a = (byte)(a + a); // OK: casting int result to byte
c = c + e; // ERROR: long -> int
c = (int)(e + f); // OK: casting long result to int
e = b + c; // OK: int -> long
y = 2 * c + e; // OK: long -> float
f = z - c * e; // ERROR: double -> long
f = (long)(z - c * e); // OK: casting double result to long
f = (long)z - c * e; // OK: casting double z to long before operation

### การแสดงข้อความ

การแสดงข้อความออกบนจอภาพ เราสามารถใช้เมทอด `System.out.print`, `System.out.println` หรือ `System.out.printf` ก็ได้ แต่ละเมทอดมีข้อแตกต่างกันดังนี้

`System.out.print` ใช้แสดงค่าค่าเดียว ค่านั้นอาจจะเป็นชนิดอะไรก็ได้

`System.out.println` ใช้แสดงค่าค่าเดียวเหมือนกับ `System.out.print` แต่ขึ้นบรรทัดใหม่หลังจากแสดงข้อความ เราสามารถเรียก `System.out.println` โดยไม่ส่งอาร์กิวเมนต์ได้เพื่อให้ขึ้นบรรทัดใหม่โดยไม่ต้องแสดงข้อความ

`System.out.printf` ใช้แสดงค่าได้หลายค่าโดยสามารถจัดรูปแบบการแสดงผลได้ การกำหนดรูปแบบการแสดงผลจะอยู่ในรูปของ format string ตามด้วยค่าที่จะแสดงผล

รูปแบบของ format string จะเป็นสตริงที่มีข้อความตามปกติ แต่สามารถจะมีตัวกำหนดรูปแบบ (format specifier) แทรกอยู่ในนั้นได้ ตัวกำหนดรูปแบบจะขึ้นต้นด้วย `%` โดยมีรูปแบบทั่วไปดังนี้

```
%[flags][width][.precision]conversion
```

- flags เป็นตัวกำหนดรูปแบบพิเศษ
- width คือจำนวนตัวอักษรขั้นต่ำที่จะแสดง ถ้าสิ่งที่จะแสดงสั้นกว่า width จะมีการเติมช่องว่างให้ครบตามที่กำหนด
- precision ในกรณีของค่าทศนิยมจะเป็นจำนวนตำแหน่งทศนิยมที่จะแสดง
- conversion คือรูปแบบที่ต้องการให้แสดง

flags, width และ precision สามารถละได้ แต่ conversion เป็นส่วนที่จำเป็นต้องระบุ

ตารางนี้แสดงตัวอย่างของตัวกำหนดรูปแบบที่มักพบได้บ่อย

| ตัวกำหนดรูปแบบ | flags | width | precision | conversion |
|--------------|-------|:-----:|:---------:|------------|
| `%d`         |       |       |           | `d` (จำนวนเต็ม) |
| `%04d`       | `0` (เติม `0` ข้างหน้าให้เต็ม) | `4`     |           | `d` |
| `%6.2f`      |       | 6     | 2         | `f` (ค่าทศนิยม) |
| `%,.1f`      | `,` (คั่นด้วย `,` ทุก 3 ตำแหน่ง) |       | 1         | `f` |
| `%s`         |       |       |           | `s` (สตริง) |
| `%n`         |       |       |           | `n` (ขึ้นบรรทัดใหม่) |
| `%%`         |       |       |           | `%` (เครื่องหมาย `%`) |

ผลลัพธ์ที่แสดงเป็นดังตัวอย่างนี้

In [77]:
System.out.printf("Just the format string and no new line... ");
System.out.printf("continuing with: %d%%%n", 10);
System.out.printf("Testing multiple values: '%04d', '%6.2f', '%,.1f'%n", 33, 4.2, 1234567.89);
System.out.printf("Testing strings: %s and %s%n", "this", "that");

Just the format string and no new line... continuing with: 10%
Testing multiple values: '0033', '  4.20', '1,234,567.9'
Testing strings: this and that


ถ้าเราไม่ได้ต้องการแสดงผล แต่ต้องการสตริงที่มีข้อความในรูปแบบที่เราต้องการ เราสามารถใช้ `String.format` ได้แทนการใช้ `System.out.printf` ข้อแตกต่างคือ `System.out.printf` จะจัดรูปแบบแล้วแสดงผล แต่ `String.format` จะจัดรูปแบบแล้วคืนค่าเป็นสตริงตามรูปแบบที่จัด

รูปแบบของ format string โดยละเอียดสามารถศึกษาได้จาก https://docs.oracle.com/javase/9/docs/api/java/util/Formatter.html

### การรับข้อมูล

เราสามารถรับข้อมูลจากผู้ใช้ผ่านทางคีย์บอร์ดได้ด้วยคลาส `Scanner`

การใช้งาน `Scanner` ต้องมีการ `import` ไลบรารีก่อนด้วยคำสั่ง

```java
import java.util.Scanner;
```

แล้วจึงสร้างอ็อบเจกต์จากคลาส `Scanner` ด้วยคำสั่ง

```java
Scanner input = new Scanner(System.in);
```

`System.in` คืออ็อบเจกต์ที่แทน standard input ซึ่งในที่นี้ก็คือคีย์บอร์ด เราตั้งชื่ออ็อบเจกต์จากคลาส `Scanner` ว่า `input` เมื่อเราได้อ็อบเจกต์แล้ว เราสามารถใช้อ็อบเจกต์นี้ในการรับค่าจากผู้ใช้ตามแบบที่เราต้องการได้โดยใช้เมทอด next และ nextLine ซึ่งคืนค่าเป็นสตริง หรือ nextInt, nextDouble, nextFloat, nextLong หรืออื่น ๆ ตามชนิดของข้อมูลที่ต้องการรับ

In [ ]:
import java.util.Scanner;

public class ScannerTest {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        System.out.print("Enter an integer: ");
        int n = input.nextInt();
        System.out.print("Enter a floating-point number: ");
        double x = input.nextDouble();
        System.out.print("Enter a few words: ");
        String firstWord = input.next();
        String rest = input.nextLine();

        System.out.println("n = " + n);
        System.out.println("x = " + x);
        System.out.println("The first word = \"" + firstWord + "\"");
        System.out.println("The rest of the line = \"" + rest + "\"");
        System.out.println("That's all!");
    }
}

เมื่อรันโปรแกรมนี้จะได้ผลลัพธ์ดังนี้

```
Enter an integer: 123
Enter a floating-point number: 1.25
Enter a few words: quick brown fox jumps     
n = 123
x = 1.25
The first word = "quick"
The rest of the line = " brown fox jumps"
That's all!
```

เมทอด `next` จะอ่านข้อความมาเป็นคำ โดยอ่านจนกว่าจะเจอช่องว่างหรือขึ้นบรรทัดใหม่ ถือเป็นอันจบคำ และคืนค่าเป็นคำที่อ่านได้ ส่วนเมทอด `nextLine` จะอ่านตั้งแต่ตำแหน่งปัจจุบันไปจนถึงขึ้นบรรทัดใหม่ และคืนค่าเป็นข้อความทั้งหมดจนถึงสิ้นสุดบรรทัด

#### ข้อควรระวังในการใช้ `Scanner`

- ชนิดของข้อมูลที่ผู้ใช้ป้อนจะต้องตรงกับชนิดที่เรารับ มิฉะนั้นจะเกิดข้อผิดพลาดในรูปของ `InputMismatchException` ซึ่งทำให้โปรแกรมหยุดทำงานได้
- การใช้ `nextLine` ต้องระมัดระวัง โดยเฉพาะการใช้ต่อจากการอ่านค่าด้วยเมทอดอื่น เนื่องจากการอ่านค่าแบบอื่นมาก่อนอาจจะทำให้ตำแหน่งปัจจุบันของข้อความเข้าอยู่ตรงท้ายบรรทัดพอดี ซึ่งเมื่ออ่านต่อด้วย `nextLine` ก็จะได้เป็นข้อความว่าง ๆ แทน ในกรณีเราสามารถแก้ไขเบื้องต้นได้ด้วยการอ่านด้วย `nextLine` 2 ครั้งต่อกัน โดยครั้งแรกจะอ่านได้ข้อความว่าง ๆ แล้วขึ้นบรรทัดใหม่ การอ่านครั้งที่สองก็จะได้ข้อความที่ต้องการจริง ๆ การแก้ไขแบบนี้อาจไม่ใช่แนวทางที่จะแก้ปัญหาได้ทุกกรณี แต่เพียงพอสำหรับการใช้งานเบื้องต้น

นอกเหนือจากการใช้ `Scanner` เราอาจจะรับข้อมูลจากผู้ใช้ได้อีกวิธีหนึ่งโดยใช้ InputStreamReader และ BufferedReader ดังตัวอย่างนี้

In [ ]:
BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
String name = reader.readLine();
System.out.println(name);

การใช้งาน `InputStreamReader` และ `BufferedReader` ต้องมีการ `import` ก่อนด้วยเช่นกัน

```java
import java.io.BufferedReader;
import java.io.InputStreamReader;
```

วิธีนี้เหมาะกับการอ่านตัวอักษรเป็นตัวหรือเป็นชุด หรืออ่านข้อความเป็นบรรทัด แต่ไม่สามารถอ่านมาเป็นตัวเลขหรือข้อมูลชนิดอื่นตรง ๆ ได้

### ธรรมเนียมในการเขียนโปรแกรมด้วยภาษา Java

ภาษาโปรแกรมแต่ละภาษาจะมีธรรมเนียมเฉพาะของภาษานั้น ๆ ที่โปรแกรมเมอร์ของภาษานั้นปฏิบัติตามกันโดยทั่วไป การเรียนรู้ภาษาใหม่ควรจะศึกษาธรรมเนียมของภาษานั้น ๆ ด้วย การไม่ยึดถือตามแนวทางปฏิบัติไม่ได้ทำให้โปรแกรมของเราใช้การไม่ได้ แต่การพัฒนาโปรแกรมร่วมกับผู้อื่นจำเป็นต้องมีแนวทางที่ถือปฏิบัติร่วมกัน มิฉะนั้นแล้วการเอาโค้ดจากโปรแกรมเมอร์แต่ละคนมาพัฒนาร่วมกันจะเป็นไปด้วยความลำบากและนำไปสู่โค้ดที่ดูยุ่งเหยิงได้

ในที่นี้เราจะเสนอธรรมเนียมที่ถือปฏิบัติกันโดยทั่วไปสำหรับโปรแกรมเมอร์ภาษา Java ได้แก่ การตั้งชื่อ การย่อหน้าในบล็อกของโค้ด

#### การตั้งชื่อ

การตั้งชื่อต่าง ๆ สามารถใข้ตัวอักษรภาษาอังกฤษทั้งตัวพิมพ์เล็กและตัวพิมพ์ใหญ่ ตัวเลข และ `_` (underscore) ประกอบกัน ตัวอักขระอีกตัวที่ใช้ได้คือ `$` แต่ตามธรรมเนียมแล้วเราจะไม่ใช้ `$` ในการตั้งชื่อทั่วไป ชื่อจะไม่สามารถขึ้นต้นด้วยตัวเลขได้

แนวทางการตั้งชื่อในภาษา Java จะใช้รูปแบบ camel case เป็นหลัก รูปแบบ camel case จะตั้งชื่อโดยเขียนทุกคำติดกันทั้งหมด ไม่แบ่งคำด้วย `_` และขึ้นต้นแต่ละคำด้วยตัวพิมพ์ใหญ่ สำหรับคำแรกอาจจะขึ้นต้นด้วยตัวพิมพ์เล็กหรือตัวพิมพ์ใหญ่ก็ได้ ถ้าขึ้นต้นด้วยตัวพิมพ์เล็กจะเรียกว่า lower camel case แต่ถ้าขึ้นต้นด้วยตัวพิมพ์ใหญ่จะเรียกว่า upper camel case

| คำต้นแบบ          | รูปแบบ camel case                   | ตัวอย่างรูปแบบที่ไม่เป็นไปตามแบบ camel case |
|------------------|------------------------------------|--------------------------------------|
| inner stopwatch  | `innerStopwatch`, `InnerStopwatch` | `innerStopWatch`, `inner_stopwatch`  |
| new customer ID  | `newCustomerId`, `NewCustomerId`   | `newCustomerID`, `new_customer_id`   |
| XML HTTP Request | `xmlHttpRequest`, `XmlHttpRequest` | `XMLHTTPRequest`, `XML_HTTP_request` |

> **หมายเหตุ** การตั้งชื่อตามแบบ camel case เป็นธรรมเนียมหรือแนวทางปฏิบัติเท่านั้น การไม่ตั้งชื่อตามธรรมเนียมไม่ได้ทำให้โปรแกรมไม่ถูกต้อง แต่เป็นสิ่งที่ควรหลีกเลี่ยง

- ชื่อตัวแปร เมทอด และพารามิเตอร์ มักจะเป็นคำนาม และใช้ lower camel case เช่น `playerStatus`, `index`
- ชื่อเมทอดมักจะเป็นคำกริยา และใช้ lower camel case เช่น `getUnusedSlots`, `processAllData`
- ชื่อคลาสมักจะเป็นคำนาม และใช้ upper camel case เช่น `GameCharacter`
- ชื่อแพกเกจจะใช้ตัวพิมพ์เล็กทั้งหมด ไม่แบ่งคำ ไม่ใช้ `_` เช่น `com.mypackage.core`
- ชื่อค่าคงที่ ซึ่งก็คือตัวแปรที่ประกาศเป็น `static final` จะใช้ตัวพิมพ์ใหญ่ทั้งหมดและแบ่งคำด้วย `_` เช่น `MAX_VALUE`, `PI`
- ชื่อตัวแปรชนิด (type variable) ซึ่งใช้กับเจเนอริกส์ (generics) จะใช้ตัวอักษรตัวพิมพ์ใหญ่ตัวเดียวและอาจตามด้วยตัวเลข เช่น `T`, `X`, `E`, `T2` หรือใช้ upper camel case ตามด้วยตัวอ้กษร `T` เช่น `RequestT`, `ValueT`

การตั้งชื่อควรตั้งให้สื่อความหมาย หลีกเลี่ยงการตั้งชื่อตัวแปรด้วยอักษรเดี่ยว ๆ ที่ไม่สื่อความหมาย ยกเว้นในกรณีที่มีความหมายเป็นที่เข้าใจโดยทั่วไป เช่น `x`, `y` ในการแทนพิกัด หรือใช้เฉพาะที่ เช่น `i`, `j` ในการวนลูป

#### การย่อหน้า

รูปแบบการย่อหน้าที่นิยมในการเขียนโปรแกรมภาษา Java มีอยู่ 2 แบบหลัก ๆ ได้แก่ แบบ K&R และแบบ Allman

ตัวอย่างแรกคือแบบ K&R ซึ่งตั้งชื่อตาม Brian Kernighan และ Dennis Ritchie ซึ่งใช้รูปแบบนี้ในภาษา C จนได้รับความนิยมอย่างกว้างขวาง

In [ ]:
public class IndentStyleDemo {
    public static void main(String[] args) {
        while (someCondition) {
            doSomething();
            doAnotherThing();
        }
        
        if (x < y) {
            doOptionOne();
        } else {
            doOptionTwo();
        }
    }
}

แบบ K&R จะเปิดวงเล็บปีกกาที่ท้ายบรรทัด ในกรณีของ `else` จะใช้วงเล็บปีกกาเปิดและปิดในบรรทัดเดียวกับ `else` เลย ในกรณีที่ในบล็อกตามหลัง `while` หรือ `if` มีเพียงคำสั่งเดียวอาจละเครื่องหมายวงเล็บปีกกาได้ รูปแบบการย่อหน้าแบบนี้เป็นแบบที่แนะนำโดย Sun Microsystems

แบบที่สองคือแบบ Allman ตามชื่อของ Eric Allman ซึ่งจะขึ้นหรือจบวงเล็บปีกกาในบรรทัดต่างหากเสมอ

In [ ]:
public class IndentStyleDemo
{
    public static void main(String[] args)
    {
        while (someCondition)
        {
            doSomething();
            doAnotherThing();
        }
        
        if (x < y)
        {
            doOptionOne();
        }
        else
        {
            doOptionTwo();
        }
    }
}

ข้อดีของแบบ Allman คือความโปร่ง เครื่องหมายเปิดปิดบล็อกวางตัวอยู่ในแนวเดียวกัน บางคนเห็นว่าแบบนี้จะดูง่ายกว่า และการทดสอบโค้ดโดยการคอมเมนต์บางส่วนออกทำได้สะดวกกว่า

การใช้รูปแบบการย่อหน้าแบบไหนก็ตาม ควรจะเป็นไปตามแบบที่ตกลงกันในทีมพัฒนา เพื่อให้โค้ดที่ได้มีความสอดคล้องกันทั้งระบบ

> **หมายเหตุ** รูปแบบการย่อหน้าแบบอื่น ๆ สามารถศึกษาเพิ่มเติมได้จาก https://en.wikipedia.org/wiki/Indentation_style

#### แท็บและช่องว่าง

การวรรคในย่อหน้าจะใช้ตัวอักษรแท็บ (tab) หรือช่องว่าง (space) การเลือกใช้แบบใดก็ตามเมื่อเลือกแล้วควรจะต้องใช้ให้เหมือนกันตลอดทั้งโค้ด

แท็บมีข้อดีที่ประหยัดตัวอักษร แต่ข้อเสียคือเวลาที่ไปแสดงบนระบบหรือเครื่องมือที่ต่างกัน จำนวนช่องว่างที่แสดงแทนแท็บอาจจะต่างกันไปด้วย ทำให้โค้ดดูต่างจากเดิม

ช่องว่างมีข้อดีที่สามารถเปิดโค้ดที่ไหนก็ได้แล้วมีการเว้นวรรคที่ดูเหมือนกันบนทุกระบบ ข้อเสียคือการขยับตำแหน่งเคอร์เซอร์ไปมาในโค้ดอาจจะใช้เวลามากกว่า

เครื่องมือสำหรับเขียนโค้ดสมัยใหม่มักจะมีตัวเลือกให้ตั้งได้ว่าเมื่อกดปุ่มแท็บ จะให้ใส่แท็บลงไปจริง ๆ หรือใส่ช่องว่างตามจำนวนที่กำหนด (เช่น 2 หรือ 4 ช่องว่าง) ลงไปแทนแท็บ

> **หมายเหตุ** ข้อแนะนำในการจัดรูปแบบโค้ดและการเขียนโค้ด Java อื่น ๆ สามารถศึกษาเพิ่มเติมได้จาก https://google.github.io/styleguide/javaguide.html